Refactor to improve `spec format` speed

tgamblin opened the pull request:
When looking at where we spend our time in solver setup, I noticed a fair bit of time is spent in `Spec.format()`, and `Spec.format()` is a pretty old, slow, convoluted method.

This PR does a number of things:
- [x] Consolidate most of what was being done manually with a character loop and several regexes into a single regex.
- [x] Precompile regexes where we keep them 
- [x] Remove the `transform=` argument to `Spec.format()` which was only used in one place in the code (modules) to uppercase env var names, but added a lot of complexity
- [x] Avoid escaping and colorizing specs unless necessary
- [x] Refactor a lot of the colorization logic to avoid unnecessary object construction
- [x] Add type hints and remove some spots in the code where we were using nonexistent arguments to `format()`.
- [x] Add trivial cases to `__str__` in `VariantMap` and `VersionList` to avoid sorting
- [x] Avoid calling `isinstance()` in the main loop of `Spec.format()`
- [x] Don't bother constructing a `string` representation for the result of `_prev_version`as it is only used for comparisons.

In my timings (on all the specs formatted in a solve of `hdf5`), this is over 2.67x faster than the original `format()`, and it seems to reduce setup time by around a second (for `hdf5`).

tgamblin committed:
spec format: simplify color formatting

This removes a bunch of old methods and makes the color formatting logic a lot clearer.

- [x] Remove `colorize_spec()`, which uses older, now incorrect colorization logic. It
      is older than the current `Spec.format()` implementation and it doesn't work with
      `@=`. The only usages in `cmd/info.py` can be replaced wtih `Spec.cformat()`.
- [x] Remove `_SEPARATORS` AND `COLOR_FORMATS`, which are no longer needed.
- [x] Simplify `Spec.format()` code to use more readable color constants directly.
- [x] Remove unused color constants.
- [x] Remove unused parameter for `write_attribute()`
- [x] Remove unused `Spec.colorized()` method

tgamblin committed:
spec format: precompile regexes and replace charater loop with `re.sub()`

tgamblin committed:
spec format: simplify parsing by using one regex

tgamblin committed:
spec format: avoid class construction during `colorize()`

Further performance improvement by avoiding object construction

`match_to_ansi` was being constructed every time. We can avoid an unnecessary object
creation by not using a class here.

tgamblin committed:
spec format: avoid StandardVersion construction in VersionList.__str__

Optimize format speed by avoiding object construction.

tgamblin committed:
spec format: avoid string comparison in `match_to_ansi`

We can get the regex engine to do more work and to avoid string comparisons.

tgamblin committed:
spec format: simplify special cases with attributes

tgamblin committed:
spec format: merge hash regexes into the spec format regex

This avoids multiple calls to parse the same string and keeps all
the parsing in one call.

tgamblin committed:
spec format: only escape and colorize if necessary

tgamblin committed:
spec format: optimize sigil check logic

tgamblin committed:
spec format: add type annotations and fix type errors

tgamblin committed:
spec format: ignore case

tgamblin committed:
spec format: handle escaped characters without negative lookbehind

We need to be able to handle escaped characters (and unescape them) anywhere in the
string. Add escaped chars to the regex and remove the now-unnecessary negative
lookbehind

tgamblin committed:
spec format: preserve behavior of original format() for '\'

The original format() method formatted '\' as '' instead of leaving it alone.
Preserve this behavior as it's relied upon by `test_spec_format_path_windows`

tgamblin commented on lib/spack/spack/spec.py:
```diff
@@ -4358,79 +4338,73 @@ def format(self, format_string=DEFAULT_FORMAT, **kwargs):
-            elif attribute.startswith("arch="):
-                sig = " arch="  # include space as separator
-                attribute = attribute[5:]
+        # TODO: This preserves a bug from the original version of this function that's needed
+        # TODO: by format_path(). Get rid of it eventually; plain '\' should just format as '\'
+        if format_string == "\\":
+            return ""
```
@scheibelp @johnwparent: this seems to be needed for `format_path`, the only place I see that formats a raw `\` string. It would be nice if it were handled in `format_path` instead of in `Spec.format()`.  Is that possible?

It's a question of whether you think a single backslash should format as `""` (old `Spec.format()` did this) or a single backslash (this `Spec.format()` would do that but for these lines). I think `\` makes more sense than this special case.

scheibelp replied:
`format_path` shouldn't be passing `\` into this method when it operates as a path separator (e.g. on Windows): it splits on path separators and formats the components individually.

Is there a test you see failing because of this?

> It's a question of whether you think a single backslash should format as "" (old Spec.format() did this) or a single backslash

I'd prefer `\` to `""`, but I'd prefer raising an exception to doing either: technically `\` is allowed in Linux paths, but it's discouraged and I don't see why we should seek out a headache (if `format_path` is letting any of those through, I'd say it's a bug, or a user is formatting `\\` in a path on Linux, which we might as well tell them not to do). 

Even though `format` is used for things other than paths, I'm not sure what allowing `\` achieves for those cases.

tgamblin replied:
> Is there a test you see failing because of this?

If you remove this check and add an assert like so:

```diff
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 520dd73f84..81bb1e33bc 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -4351,8 +4351,8 @@ def format(self, format_string: str = DEFAULT_FORMAT, color: Optional[bool] = Fa

         # TODO: This preserves a bug from the original version of this function that's needed
         # TODO: by format_path(). Get rid of it eventually; plain '\' should just format as '\'
-        if format_string == "\\":
-            return ""
+        #        if format_string == "\\":
+        #            return ""

         def safe_color(sigil: str, string: str, color_fmt: Optional[str]):
             # avoid colorizing if there is no color or the string is empty
@@ -4513,6 +4513,9 @@ def format_path(
         else:
             output_path_components = []
             input_path_components = list(format_string_as_path.parts)
+
+        assert input_path_components[0] != "\\"
+
         output_path_components += [
             fs.polite_filename(self.format(x)) for x in input_path_components
         ]
```

then run:

```console
$ spack unit-test --showlocals -vv 'lib/spack/spack/test/spec_semantics.py::test_spec_format_path_windows'
```

You get:

```console
lib/spack/spack/test/spec_semantics.py:1128: in _check_spec_format_path
    formatted = spec.format_path(format_str, _path_ctor=path_ctor)
        expected   = 'installroot\\git-test\\git.foo_bar'
        format_str = '/installroot/{name}/{version}'
        path_ctor  = <class 'pathlib.PureWindowsPath'>
        spec       = git-test@git.foo/bar
        spec_str   = 'git-test@git.foo/bar'
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = git-test@git.foo/bar, format_string = '/installroot/{name}/{version}'
_path_ctor = <class 'pathlib.PureWindowsPath'>

    def format_path(
        # self, format_string: str, _path_ctor: Optional[pathlib.PurePath] = None
        self,
        format_string: str,
        _path_ctor: Optional[Callable[[Any], pathlib.PurePath]] = None,
    ) -> str:
        """Given a `format_string` that is intended as a path, generate a string
        like from `Spec.format`, but eliminate extra path separators introduced by
        formatting of Spec properties.

        Path separators explicitly added to the string are preserved, so for example
        "{name}/{version}" would generate a directory based on the Spec's name, and
        a subdirectory based on its version; this function guarantees though that
        the resulting string would only have two directories (i.e. that if under
        normal circumstances that `str(Spec.version)` would contain a path
        separator, it would not in this case).
        """
        format_component_with_sep = r"\{[^}]*[/\\][^}]*}"
        if re.search(format_component_with_sep, format_string):
            raise SpecFormatPathError(
                f"Invalid path format string: cannot contain {{/...}}\n\t{format_string}"
            )

        path_ctor = _path_ctor or pathlib.PurePath
        format_string_as_path = path_ctor(format_string)
        if format_string_as_path.is_absolute():
            output_path_components = [format_string_as_path.parts[0]]
            input_path_components = list(format_string_as_path.parts[1:])
        else:
            output_path_components = []
            input_path_components = list(format_string_as_path.parts)

>       assert input_path_components[0] != "\\"
E       AssertionError

_path_ctor = <class 'pathlib.PureWindowsPath'>
format_component_with_sep = '\\{[^}]*[/\\\\][^}]*}'
format_string = '/installroot/{name}/{version}'
format_string_as_path = PureWindowsPath('/installroot/{name}/{version}')
input_path_components = ['\\', 'installroot', '{name}', '{version}']
output_path_components = []
path_ctor  = <class 'pathlib.PureWindowsPath'>
self       = git-test@git.foo/bar
```

Without the assert, the test fails like this:

```console
lib/spack/spack/test/spec_semantics.py:1155:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

spec_str = 'git-test@git.foo/bar', format_str = '/installroot/{name}/{version}'
expected = 'installroot\\git-test\\git.foo_bar', path_ctor = <class 'pathlib.PureWindowsPath'>

    def _check_spec_format_path(spec_str, format_str, expected, path_ctor=None):
        spec = Spec(spec_str)
        if not expected:
            with pytest.raises((spack.spec.SpecFormatPathError, spack.spec.SpecFormatStringError)):
                spec.format_path(format_str, _path_ctor=path_ctor)
        else:
            formatted = spec.format_path(format_str, _path_ctor=path_ctor)
>           assert formatted == expected
E           AssertionError: assert '_\\installroot\\git-test\\git.foo_bar' == 'installroot\\git-test\\git.foo_bar'
E
E             - installroot\git-test\git.foo_bar
E             + _\installroot\git-test\git.foo_bar
E             ? ++

expected   = 'installroot\\git-test\\git.foo_bar'
format_str = '/installroot/{name}/{version}'
formatted  = '_\\installroot\\git-test\\git.foo_bar'
path_ctor  = <class 'pathlib.PureWindowsPath'>
spec       = git-test@git.foo/bar
spec_str   = 'git-test@git.foo/bar'
```

scheibelp replied:
That is specifically for the 

`r"/installroot/{name}/{version}"`

test case on windows, which is a nonsense path. As long as it formats to something we expect, the output can be anything (doesn't have to be a valid path)

In other words, it's fine to set `expected` to `_\installroot\git-test\git.foo_bar` in this case (and remove these lines):

(`['\\', 'installroot', '{name}', '{version}']` are the input path components, and `polite_filename` makes sure `\\` is transformed into `_` so it has no significance as a directory)

Does that seem agreeable?

tgamblin replied:
So personally I think it would be more consistent if `format_path` noticed this:

```
input_path_components = ['\\', 'installroot', '{name}', '{version}']
```

And just removed the leading backslash if it is truly irrelevant on windows. If what you say is true (`format_path` should not be passing a single `\` to format()) then the behavior seems way better to me, and the expected result is not dependent on what is essentially a corner case for `format()`: an escape character by itself.  I can add that to this PR if you think it makes sense.

I can also make it so that `format()` raises with a single `\`, which is basically what Python does (it raises for an unterminated string literal if you try to get it to parse that).

tgamblin replied:
This does make me wonder, though, whether `format()` shouldn't just behave like Python and use `{{` and `}}` as its escape characters instead of `\{` and `\}`.  Would that eliminate the need to handle path formatting specially?  In a lot of ways I wonder why we need `format_path` and whether anyone's actually escaping their `{` or `}` characters in spack format strings.

We don't actually use the escaping anywhere in the Spack code base.

tgamblin replied:
ok, after looking at the code a bit, I decided to do two things:
1. raise an error in `Spec.format()` if it's passed an unterminated escape, i.e. `'\'` or `'foobar\'`
2. preserve any leading `'/'` as `'\'`, even if it's inconsequential. This is what `PureWindowsPath` does, so I think it's the most reasonable.

We could probably eliminate the need for `format_path` by using double-brace escapes but that can be some other PR.

tgamblin replied:
@scheibelp does this work for you?

scheibelp replied:
> We could probably eliminate the need for format_path by using double-brace escapes but that can be some other PR.

`format_path` isn't about what goes in: changing from `{` to `{{` would not fix what it is made for. `format_path` makes sure that Spec attributes like `Spec.version` don't generate characters that either (a) look like path separators or (b) look like other potentially problematic characters (e.g. `=` in makefiles).

At least, I don't see how switching to `{{` would address that, so I could use some more explanation to understand what you mean (not here necessarily, since you didn't make that change here, but whenever you intend to do it).

> raise an error in Spec.format() if it's passed an unterminated escape, i.e. '\' or 'foobar\'

Looking at https://github.com/spack/spack/pull/43712/commits/14c8f70358d2b74ba9870e9eed133430eb6ab553 I'm a bit worried about the complexity of SPEC_FORMAT_RE (and e.g. have trouble seeing what happens if an unterminated escape appears somewhere besides the end of the string), but the overall desired behavior sounds good to me.

> preserve any leading '/' as '\', even if it's inconsequential. This is what PureWindowsPath does, so I think it's the most reasonable.

Users wouldn't want the result regardless, so it all seems fine to me (including what you chose), although the way you achieved this seems somewhat roundabout:

```
        output_path_components += [
            fs.polite_filename(self.format(x)) for x in input_path_components
            fs.polite_filename(self.format(part)) if part != "\\" else part
            for part in input_path_components
        ]
```

this logic appears to consider cases of `\\` that are not just at the beginning: if we only intend to deal with leading `\\` in this way, I think this code should be explicit about it.

I'd recommend a comment there about how a '\\` can even come up, given that the string was just fed to something that "should" remove these.

tgamblin replied:
> `format_path` isn't about what goes in: changing from `{` to `{{` would not fix what it is made for. `format_path` makes sure that Spec attributes like `Spec.version` don't generate characters that either (a) look like path separators or (b) look like other potentially problematic characters (e.g. `=` in makefiles).

Ok, sorry I wasn't clear. It would eliminate the need to split the path by separator, but I think it has its own set of special cases, and it's not critical. I don't think it is a particularly needed change.

>  I'm a bit worried about the complexity of SPEC_FORMAT_RE

I don't think we can avoid this -- that regex is the whole reason that this is fast.

> e.g. have trouble seeing what happens if an unterminated escape appears somewhere besides the end of the string

That doesn't happen. If it occurs in the string option 1 consumes it and the character after `\` is output. The only way option 1 doesn't match is if there is a `\` at the end of the string; otherwise there is always something after the `\`.

> the overall desired behavior sounds good to me.

that's good!

> although the way you achieved this seems somewhat roundabout:
> 
> ```
>         output_path_components += [
>             fs.polite_filename(self.format(x)) for x in input_path_components
>             fs.polite_filename(self.format(part)) if part != "\\" else part
>             for part in input_path_components
>         ]
> ```
> 
> this logic appears to consider cases of `\\` that are not just at the beginning: if we only intend to deal with leading `\\` in this way, I think this code should be explicit about it.
> 
> I'd recommend a comment there about how a '\\` can even come up, given that the string was just fed to something that "should" remove these.

I investigated this further. As you say, the *only* place you can have a `\\` *on Windows* is in fact the beginning of the string. I was thinking you could get it somewhere else, but almost any other place I put backslashes they're ignored or treated as path separators. I think there are three cases on Windows:

* Initial `C:\` (or other letter) is absolute and results from a path like `C:\\foo/bar/baz`
* Initial `\\server\c\` is a remote server path and volume. It's interpreted as absolute and results from a path like `\\server\c\foo\bar\baz`
* Initial `\` happens when you pass a path that begins with `\` or `/`

The first two are handled by the existing logic; the last one is not.  However, the last one can occur on Posix.  In fact, any path component can end with `\\` on Posix, which will break `format()`:

```
>>> Path(r"\/foo\/bar/baz").parts
('\\', 'foo\\', 'bar', 'baz')
```

So the handling actually needs to be different here. Any path component can legitimately end with an unterminated `\\`.  Yes it's unlikely but it is allowed.

I think based on this, the reasonable thing to do for a final `\\` is *not* to raise but to just have `format()` leave the `\\` alone (i.e. just preserve it).  That would solve this in one place and would simplify `SPEC_FORMAT_RE` as well as the logic in `format_path`.  I can do that and add a comment. Agree?

scheibelp replied:
> Ok, sorry I wasn't clear. It would eliminate the need to split the path by separator, but I think it has its own set of special cases, and it's not critical. I don't think it is a particularly needed change.

I still don't 100% follow (but again, that doesn't worry me here): `format_path` not only splits on path separators, but also calls `polite_filename` on each component.

> I think based on this, the reasonable thing to do for a final \\ is not to raise but to just have format() leave the \\ alone (i.e. just preserve it). That would solve this in one place and would simplify SPEC_FORMAT_RE as well as the logic in format_path. I can do that and add a comment. Agree?

I think this means you are talking about modifying the Windows test like described in https://github.com/spack/spack/pull/43712#discussion_r1571872941 - if I understood that correctly, then I agree

tgamblin replied:
Per discussion with Peter we decided to:
* Not raise for an unterminated `\` in format strings (i.e., leave it untouched)
* On windows, pass an initial `\` through as we would for an absolute path, so that the result of `format_path()` is consistent with `PureWindowsPath` on Windows.
* On posix, treat an initial `\` as a path component (as it must be on posix) and format it.r

This simplifies the ragex and the logic in `format_path`, and it remains consistent with how `PureWindowsPath` behaves.

tgamblin committed:
use `typing.Match` instead of `re.Match`

haampie commented:
> Note that this does not memoize the result of Spec.format(), as @haampie had suggested to do

tgamblin commented:
If you want to try this out, I made a little benchmark here: https://github.com/tgamblin/spec-format-perf

tgamblin commented:
Ok that makes more sense. I think actually memoizing the format result during setup is probably going to be a bigger win.

tgamblin committed:
`fake_dag_hash` should take a length argument

tgamblin committed:
handle `tty.msg(None)`

tgamblin committed:
spec format: fail on bare '\' character

- [x] raise an exception for an unterminated escape character ('\')
- [x] format_path preserves leading '/' as '\' when on windows

tgamblin committed:
spec format: avoid an extra `str()` in `cescape()`

Avoid a constructor call if it's not needed.

tgamblin committed:
spec format: remove unnecessary `transform=` argument

`transform=` adds a whole lot of extra complexity to `Spec.format()`, and the thing
it was originally intended for (the `decorator=` arg to `spack.cmd.display_specs()`) is
now handled entirely in `display_specs()`.

The only place it is still used is in the common modules code, and we can handle what
it's used for there with a simple `str.upper()`. If people want to format different
components of specs differently, they can call `format()` multiple times.

tgamblin commented:
@alalazo can you look at [30d85cb](https://github.com/spack/spack/pull/43712/commits/30d85cb11cd573fc496549f242853536b64b4ac6) (`spec format: remove unnecessary transform= argument`) and make sure you agree?

tgamblin commented:
As of now I'm getting:
* 1.155e-4 s/format (this), vs
* 2.655e-4 s/format (develop)

So this is ~2.3x faster.

tgamblin committed:
spec format: a few more optimizations

- [x] trivial cases for VariantMap and VersionList
- [x] don't bother making a string in `_prev_version` (which is used only in comparisons)

tgamblin committed:
spec format: avoid most `isinstance()` calls with name check

tgamblin commented:
Ok now it's 2.59x faster.  I think I have wrung all I can out of this. If we can avoid somehow constructing a `StandardVersion` in `_prev_version`, it might help a bit more, but I do not see an easy way to do that.

I think this is ready to go

tgamblin committed:
spec format: fix windows path handling per conversation with Peter

alalazo commented on lib/spack/llnl/util/tty/__init__.py:
```diff
@@ -162,7 +162,9 @@ def msg(message, *args, **kwargs):
 
-    if isinstance(message, Exception):
+    if message is None:
+        message = "None"
```
Why is this necessary? The message shouldn't be optional, or maybe we should consider this a no-op. I don't think we expect this to happen:
```console
Spack version 0.22.0.dev0
Python 3.11.5, Linux x86_64
>>> from llnl.util import tty
>>> tty.msg(None)
==> None

```

alalazo replied:
We should also mark this file and have a subsequent PR removing all `*args, **kwargs` :slightly_smiling_face:

tgamblin replied:
`tty.msg` does this on `develop`, so I'm just preserving the behavior.  I cleaned this up and added type annotations to `tty.msg` but I'm preserving this behavior -- at least one test fails if you don't handle `None`.  I think we can clean that up later.

alalazo commented on lib/spack/llnl/util/tty/color.py:
```diff
@@ -305,7 +293,7 @@ def cprint(string, stream=None, color=None):
 
 
-def cescape(string):
+def cescape(string: str):
```
```suggestion
def cescape(string: str) -> str:
```

tgamblin replied:
done

alalazo commented on lib/spack/spack/cmd/info.py:
```diff
@@ -263,8 +263,8 @@ def _fmt_name_and_default(variant):
 
-def _fmt_when(when, indent):
-    return color.colorize(f"{indent * ' '}@B{{when}} {color.cescape(when)}")
+def _fmt_when(when: bool, indent: int):
```
Is `when` a bool or a `Spec`?

tgamblin replied:
fixed

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -4361,79 +4346,68 @@ def format(self, format_string=DEFAULT_FORMAT, **kwargs):
-            transform (dict): maps full-string formats to a callable \
-                that accepts a string and returns another one
-
+            color: True if returned string is colored, False if not,
```
We can delete `Keyword Args:` above

tgamblin replied:
done

alalazo commented on lib/spack/spack/spec.py:
```diff
@@ -4361,79 +4346,68 @@ def format(self, format_string=DEFAULT_FORMAT, **kwargs):
-                attribute = attribute[1:]
-                dep, attribute = attribute.split(".", 1)
-                current = self[dep]
+        def safe_color(sigil: str, string: str, color_fmt: Optional[str]):
```
Missing return type in type hint

alalazo commented on lib/spack/spack/test/spec_semantics.py:
```diff
@@ -703,22 +703,25 @@ def check_prop(check_spec, fmt_str, prop, getter):
-
-        sigil_mismatches = [
+    @pytest.mark.parametrize(
+        "fmt_str",
+        [
             "{@name}",
             "{@version.concrete}",
             "{%compiler.version}",
             "{/hashd}",
             "{arch=architecture.os}",
-        ]
+        ],
+    )
```
:+1:

alalazo commented on lib/spack/spack/variant.py:
```diff
@@ -638,6 +638,9 @@ def copy(self):
         return clone
 
     def __str__(self):
+        if not self:
+            return ""
```
Where does this happen? Do we call `VariantMap.__str__` explicitly anywhere?

tgamblin replied:
it's called by `format()` if you format with `"{variants}"`, which is in the standard format string.  This optimizes the (very) common case of having no variants.  e.g., most abstract specs in package metadata don't have variants.

alalazo commented on lib/spack/spack/version/version_types.py:
```diff
@@ -1019,6 +1024,9 @@ def __hash__(self):
         return hash(tuple(self.versions))
 
     def __str__(self):
+        if not self.versions:
+            return ""
```
Why is this one also necessary? Looking at the class, versions could be an empty list - and the expression below returns `""` in that case. Am I missing some corner case?

tgamblin replied:
This is another fast path for the common case (like the one for variants)

alalazo replied:
:thinking: I don't understand how this is a fast path, but it doesn't harm.

alalazo reviewed (commented):
I just have a few comments. This commit https://github.com/spack/spack/pull/43712/commits/0771acacb8679755add96cfde1731f3e9825fdc1 LGTM

tgamblin committed:
spec format: clean up `tty.msg`

tgamblin committed:
spec format: add return type to cescape()

tgamblin committed:
fix type in _fmt_when

tgamblin committed:
spec format: fix format() docstring

tgamblin committed:
spec format: return type for safe_color
